parlai/core/pytorch_data_teacher.py

```diff
@@ -385,31 +386,57 @@ def __init__(self, opt, shared=None):
         self.num_workers = opt['numworkers']
         self.batch_cache_type = opt.get('batch_sort_cache')
         # One can specify a collate function to use for preparing a batch
-        collate_fn = opt.get('collate_fn', default_collate)
+        self.opt = copy.deepcopy(opt)
+        dataset_class, self.collate_fn = self.get_dataset_class(opt)
+        opt['dataset'] = dataset_class
+        opt['collate_fn'] = self.collate_fn
+
         if not shared:
-            self.dataset = StreamDataset(opt)
+            self.dataset = dataset_class(opt)
             self.pytorch_dataloader = DataLoader(
                 self.dataset,
                 batch_size=self.bsz,
                 shuffle=False,
                 sampler=sampler.SequentialSampler(self.dataset),
                 num_workers=self.num_workers,
-                collate_fn=collate_fn,
+                collate_fn=self.collate_fn,
                 pin_memory=False,
                 drop_last=False,
                 )
             self.lastYs = [None] * self.bsz
             if self.batch_cache_type != 'none':
                 self.loader_process = LoaderProcess(opt)
                 self.loader_process.start()
+            self.data = enumerate(self.pytorch_dataloader)
         else:
             self.dataset = shared['dataset']
             self.pytorch_dataloader = shared['pytorch_dataloader']
             self.lastYs = shared['lastYs']
+            self.data = shared['data']
 
         self.num_batches = math.ceil(self.dataset.num_episodes()/self.bsz)
         self.reset()
 
+    def get_dataset_class(self, opt):
+        dataset_name = opt.get('dataset')
```

Comment:
Can you add a docstring here? Good to provide a few examples of how to specify the dataset name

Reply:
sure!